geom/type_geometry_collection.go

```diff
@@ -229,8 +229,140 @@ func (c GeometryCollection) Area() float64 {
 	var sum float64
 	n := c.NumGeometries()
 	for i := 0; i < n; i++ {
-		geom := c.GeometryN(i)
-		sum += geom.Area()
+		g := c.GeometryN(i)
+		sum += g.Area()
 	}
 	return sum
 }
+
+func highestDimensionIgnoreEmpties(g Geometry) int {
+	// We could simply use Dimension() instead of this function
+	// except for the fact empties can have different dimensionalities.
+	// This function thus exists to treat empties as dimensionality 0.
+	if g.IsEmpty() {
+		return 0
+	}
+	if !g.IsGeometryCollection() {
+		return g.Dimension()
+	}
+	c := g.AsGeometryCollection()
+	highestDim := 0
+	for _, g2 := range c.geoms {
+		highestDim = max(highestDim, highestDimensionIgnoreEmpties(g2))
+	}
+	return highestDim
+}
+
+// Centroid of a GeometryCollection is the centroid of its parts' centroids.
+// It returns true iff the centroid is well defined.
+func (c GeometryCollection) Centroid() (Point, bool) {
+	result, valid := c.sumCentroidCalc()
+	var xy XY
+	switch result.highestDim {
+	case 0:
+		if result.numPoints == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 0 and numPoints is 0
+		}
+		xy = result.sumXY.Scale(1.0 / float64(result.numPoints))
+	case 1:
+		if result.sumLength == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 1 and sumLength is 0
+		}
+		xy = result.sumXY.Scale(1.0 / result.sumLength)
+	case 2:
+		if result.sumArea == 0 {
+			return Point{}, false // Invalid centroid, highestDim is 2 and sumArea is 0
+		}
+		xy = result.sumXY.Scale(1.0 / result.sumArea)
+	default:
+		panic("Invalid dimensionality in centroid calculation.")
+	}
+
+	return NewPointXY(xy), valid
+}
+
+type centroidCalc struct {
+	highestDim int
+	numPoints  int
+	sumLength  float64
+	sumArea    float64
+	sumXY      XY
+}
+
+func (c GeometryCollection) sumCentroidCalc() (result centroidCalc, valid bool) {
```

Comment:
I'm not quite sure I understand why we need the `valid` flag. It seems as though an invalid centroid calculation could just be encoded naturally inside the `centroidCalc` (e.g. when the highest dim is 1, then a result of `centroidCalc{highestDim: 1, sumLength: 0}` would represent an invalid calculation).

Reply:
Remove validity flag.